Hacky Hour is back!

3 themes this time!

## Workflow improvement
To solve [issue #31](https://github.com/ScienceParkStudyGroup/studyGroup/issues/31) submitted by Grace (PhD student, SILS Molecular Neuroscience).

Quoting: 
> I have looking into the sequence conservation across a number of gene promoters. I am interested specifically in the indels and substitutions. To do this analysis I joined together multiple sequences (from the same species) into one long sequence with 'spacer' sequences between and aligned this with an amalgamated sequence from another species as one alignment. Using this alignment I can then either split it up again into the individual promoter sequences or 'code' it based on substitutions/indels to create a sort of heatmap.
Currently I am having difficulty pulling out the alignments that I am interested in based on certain critera. I am also aware that my workflow is incredibly unsophisticated and messy. Currently I start at UCSC, move to bash, to R and then to excel. It would be amazing to get some input into what I can do to clean up my pipeline and write some slightly more sophisticated code!
Thanks,
Grace

## Identifying potential gene regulatory networks
To solve [issue #32](https://github.com/ScienceParkStudyGroup/studyGroup/issues/32) submitted by Fred (Master student, SILS Molecular Neuroscience).

Quoting: 
> My current problem is trying to identify potential gene regulatory networks involving transposable elements (TEs) using 10 samples and rnaseq count data. Currently I have performed a correlational analysis on elements that are within 50kb of a gene. This is good as a preliminary analysis however due to the small sample size (n=10) this is not a very robust way of looking at the data.
My current dataset contains all loci, the distance between a TE and a gene (if within 50kb) and correlation coefficients as well as p-values.
I also have raw count data.
What's an intelligent way to proceed to potentially identify some TEs and their related genes?
If clarification is needed then let me know!
Cheers
Fred

## Needed Shiny application for lab experiments
To see what type of Shiny application Bas Beerens (SILS, Molecular Plant Phytopathology) would need.

## Announced helpers:
- Huub Hoefsloot (SILS, Biosystems & Data Analysis)
- Stefania Astrologo (SILS, Synthetic biology & nuclear organisation)
- Marc Galland (SILS, Plant Physiology)
