# Metabarcoding for Microbial Ecology

 - **Authors**: 
   - Susanne Wilken (Assistant-Professor, UvA-IBED FAME)
   - Mattias de Hollander (Bioinformatician NIOO)
   - Marc Galland (Postdoc, UvA-SILS)
 - **Research field**: Microbial Ecology / Bioinformatics
 - **Lesson topic**: metabarcoding through 16S and 18S rRNA  
 - **Lesson content URL**: <https://github.com/ScienceParkStudyGroup/studyGroup/tree/gh-pages/lessons/20180626_Metabarcoding_Susanne/>

# Introduction


# Lesson objectives
At the end of this lesson, you will:
- Be able to understand the major steps of a 16S/18S rRNA metabarcoding analysis
- Have undertaken one to several steps of the workflow (see below)
- Have identified potential pitfalls in these steps.

# Analysis steps
- **Step 1**: Quality filter raw reads (example workflow for workshop)  
- **Step 2**: Dereplicate your dataset    
- **Step 3**: Make a sequence variant table  
- **Step 4**: Identify chimeras and remove chimeric sequences   
- **Step 5**: Assign taxonomy to your sequences based on a reference database    
- **Step 6**: Visualize and analyze microbial community composition  
 

# Going further
- [Callahan et al., 2018](https://www.nature.com/articles/nmeth.3869)
- [Silva ribosomal RNA databases](https://www.arb-silva.de/) 
