# Metabarcoding for Microbial Ecology

 - **Authors**: 
   - Susanne Wilkens (Assistant-Professor, UvA-IBED FAME)
   - Mattias de Hollander (Bioinformatician NIOO)
   - Marc Galland (Postdoc, UvA-SILS)
 - **Research field**: Microbial Ecology / Bioinformatics
 - **Lesson topic**: 16S rRNA  
 - **Lesson content URL**: <https://github.com/ScienceParkStudyGroup/studyGroup/tree/gh-pages/lessons/20180626_Metabarcoding_Susanne/>

# Introduction


# Lesson objectives
At the end of this lesson, you will:
- Be able to understand the major steps of a 16S rRNA metabarcoding analysis
- Have undertaken one to several steps of the workflow (see below)
- Have identified potential pitfalls in these steps.

# Steps
**Step 1**: Demultiplex the raw reads (example workflow for workshop)  
**Step 2**: Pick Operational Taxonomic Units (OTUs)  
**Step 3**: Identify chimeras and remove chimeric sequences from the OTU table  
**Step 4**: Align sequences and remove alignment failures from the OTU table  
**Step 5**: Filter rep set fasta file to match the OTU IDs in your filtered OTU table  
**Step 6**: Construct a phylogenetic tree  
**Step 7**: Carry out microbial community analyses to assess alpha- and beta-diversity  
 
