{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prerequisites\n",
    "[MUMmer](http://mummer.sourceforge.net/)  MUMmer does not run on Windows systems unfortunately. As an alternative, you can use Mauve to align genomes, this has a slightly different output format than MUMmer. \n",
    "[Circos](http://circos.ca/) \n",
    "Programs will be run from the commandline, so you need to be able to open a terminal. Google `terminal` + the name of your operating system, and you'll fiind out how to do it. On a Mac it is in /Applications/Utilities/. In this document, command are written in unix form (Linux/MacOS). Most will work in widnows terminal too, just flip the slashes (change / into \\). If not, check out a cheatsheet (e.g. [this one](http://ftp.kh.edu.tw/Linux/Redhat/en_6.2/doc/gsg/ch-doslinux.htm) for 'translations' )\n",
    "\n",
    "##  Compare genomes of closley related species  \n",
    "\n",
    "Two relatively closely related pathogenic fungi, _Fusarium oxysporum_ (_Fo_) and _Fusarium graminearum_ (_Fg_) have a very different chromosome number. _Fg_ has 4 chromosomes, while _Fo_ has 15. They also differ in size: _Fg_ has a genome of 36.36 Mb (megabases: 1 million A,T,C or G) and _Fo_ has 57.91 Mb. We want to compare these two genomes and see whether for example one chromsoome in _Fg_ corresponds to two or more chromosomes in _Fo_.\n",
    "\n",
    "For this we need to download and align the genomes of _Fo_ and _Fg_. We will use Circos to visualize these alignments. \n",
    "\n",
    "  \n",
    "### 0. Project management\n",
    "Let's first do a little project management, and make a directory structure in which we will save the data we download and the results of our analyses.\n",
    "\n",
    "For this, open a Terminal and create a working directory, something like:  \n",
    "`mkdir CircosPlot_FgFo`  \n",
    "`mkdir` is the command for creating a directory.  \n",
    "  \n",
    "This notebook is run from this directory, that is I used `cd` to `c`hange `d`irectory to my working directory `CircosPlot_FgFo`, and used `jupyter notebook` to start this notebook.  \n",
    "\n",
    "Create a folder for data:  \n",
    "`mkdir data/`  \n",
    "We may add other data types to the circos plot (Circos is very suitable for that) so let's make a separate subdirectory for genomesequences:  \n",
    "`mkdir data/genomes/` \n",
    "  \n",
    "  \n",
    "We also need a folder to store the results of our comparative analyses:  \n",
    "`mkdir results/`\n",
    "We need a place to store the results from genome comparisons:  \n",
    "`mkdir results/genome_alignments/` \n",
    "  \n",
    "We will visualize these results in a single plot, using Circos:  \n",
    "`mkdir circos/`\n",
    "Circos needs a number of input files, which we will store in a separate directory:  \n",
    "`mkdir circos/input/`\n",
    "Circos needs config files, files in which you state the configuration of your plot, i.e. how you plot will look.  \n",
    "`mkdir circos/input/config_files/`  \n",
    "Circos needs files describing the chromsomes you want to plot:  \n",
    "`mkdir circos/input/karyotypes/`  \n",
    "and the links:  \n",
    "`mkdir circos/input/links/`  \n",
    "   \n",
    "Circos output we will store in:\n",
    "`mkdir circos/output/`  \n",
    "  \n",
    "  \n",
    "\n",
    "It is also good practice to redirect any output (like error messages, progress reports, etc. to a logfile. Let's save these in a separate directory as well:\n",
    "`mkdir docs/`\n",
    "\n",
    "We work from the commandline mostly, and it is vital you save the commands you've typed for future reference. This is like your labjournal. Take it seriously! \n",
    "\n",
    "It is also good practice to write some information in the project and the analyses you've done and where you stored what in a README file. For example a description of the directory structure, etc. as I have given above.  \n",
    "\n",
    "One way to combine all this is by using a notebook, like this one.  \n",
    "  \n",
    "\n",
    "  \n",
    "### 1. Download genomes of _Fusarium oxysporum_ and _Fusarium graminearum_\n",
    "\n",
    "Go to the [genbank website](https://www.ncbi.nlm.nih.gov/).  \n",
    "From `All Databases` (on the left), select `Assembly` and type `Fusarium oxysporum f. sp. lycopersici 4287` in the search bar [link-to-result](https://www.ncbi.nlm.nih.gov/assembly/?term=Fusarium+oxysporum+f.+sp.+lycopersici+4287).  \n",
    "You can choose from three different assemblies. Pick the one with `Assembly level: Chromosome` [ASM14995v2](https://www.ncbi.nlm.nih.gov/assembly/GCF_000149955.1).\n",
    "\n",
    "Click on the [link to the assembly](https://www.ncbi.nlm.nih.gov/nuccore/AAXH00000000.1/), scroll down, click on the chromosome sequences (accession starts with CM): [link](https://www.ncbi.nlm.nih.gov/nuccore?term=CM000589:CM000603[PACC]). From the menu on the right `Send to:` select `file`, format `FASTA`, sort by `Accession`. You will download a file called `sequence.fasta.txt`, rename this file (`mv` (`m`o`v`e) in the terminal) to `Fol4287.fasta`. Store it in `data/genomes/` (e.g. `mv Downloads/sequence.fasta.txt CircosPlot_FgFo/data/genomes/Fol4287.fasta`)  \n",
    "\n",
    "Check the contents of this file using e.g `grep '>' ~home/CircosPlot_FgFo/data/genomes/Fol4287.fasta`.  You should get:\n",
    "\n",
    "\\>CM000589.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 1, whole genome shotgun sequence  \n",
    "\\>CM000590.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 2, whole genome shotgun sequence  \n",
    "\\>CM000591.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 3, whole genome shotgun sequence  \n",
    "\\>CM000592.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 4, whole genome shotgun sequence  \n",
    "\\>CM000593.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 5, whole genome shotgun sequence  \n",
    "\\>CM000594.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 6, whole genome shotgun sequence  \n",
    "\\>CM000595.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 7, whole genome shotgun sequence  \n",
    "\\>CM000596.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 8, whole genome shotgun sequence  \n",
    "\\>CM000597.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 9, whole genome shotgun sequence  \n",
    "\\>CM000598.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 10, whole genome shotgun sequence  \n",
    "\\>CM000599.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 11, whole genome shotgun sequence  \n",
    "\\>CM000600.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 12, whole genome shotgun sequence  \n",
    "\\>CM000601.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 13, whole genome shotgun sequence  \n",
    "\\>CM000602.2 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 14, whole genome shotgun sequence  \n",
    "\\>CM000603.1 Fusarium oxysporum f. sp. lycopersici 4287 chromosome 15, whole genome shotgun sequence   \n",
    "\n",
    "Note that chromosome 14 ends with `.2`. This is because the assembly was updated and has a second version for this chromosome.\n",
    "\n",
    "Now do the same for F. graminearum: (pick the one with accession GCF_000240135.3, sequenced and assembled by the Broad institute)\n",
    "(https://www.ncbi.nlm.nih.gov/assembly/GCF_000240135.3/)\n",
    "WGS project, select accession starting with CM, send to file, etc, etc. \n",
    "Rename the file to `Fg_PH1.fasta`. Store it in `CircosPlot_FgFo/data/genomes/` \n",
    "\n",
    "`grep '>' Fg_PH1.fasta`.  \n",
    "\\>CM000574.1 Fusarium graminearum PH-1 chromosome 1, whole genome shotgun sequence  \n",
    "\\>CM000575.1 Fusarium graminearum PH-1 chromosome 2, whole genome shotgun sequence  \n",
    "\\>CM000576.1 Fusarium graminearum PH-1 chromosome 3, whole genome shotgun sequence  \n",
    "\\>CM000577.1 Fusarium graminearum PH-1 chromosome 4, whole genome shotgun sequence  \n",
    "\n",
    "\n",
    "  \n",
    "### 2. Align the genomes of _Fusarium oxysporum_ and _Fusarium graminearum_  \n",
    "\n",
    "Next, we want to compare these genomes, to see which genomic regions are homologous to eachother. For this, we will use [MUMmer](http://mummer.sourceforge.net/). MUMmer is a package that contains programs to align genomes, filter aligned regions and plot algnments in dotplots. MUMmer is well-documented.\n",
    "\n",
    "We will use promer to align _Fo_ and _Fg_. Promer first translates dna sequences into protein and then aligns those. It is more sensitive over longer evolutionary distances than nucmer (that aligns DNA sequences directly).  \n",
    "\n",
    "I installed MUMmer version 3.23 in my /Applications folder so I will run the genome alignment like this:\n",
    "`/Applications/MUMmer3.23/promer`  \n",
    "I will add prefix (using `-p`) to my output files, so they will be saved in the correct directory and they will start with this name. As I align Fol4287 with Fg_PH1 using promer with default settings, I choose:\n",
    "`-p results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def`\n",
    "Then we need to tell promer which genomes to align:\n",
    "`~data/genomes/Fol4287.fasta` and `data/genomes/FG_PH1.fasta`.  \n",
    "and we'll redirect (`>&`: `>` is to redirect output that would normally be printed in the terminal (also known as `stdout`, I add `&` to also include errormessages) the output of this to a logfile, which we will store in `docs`:\n",
    "`docs/Fol4287__vs__Fg_PH1.promer_def.log`.  \n",
    "\n",
    "The entire command is:\n",
    "`/Applications/MUMmer3.23/promer -p results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def data/genomes/Fol4287.fasta data/genomes/FG_PH1.fasta >& docs/Fol4287__vs__Fg_PH1.promer_def.log`.\n",
    "\n",
    "This takes a few minutes...\n",
    "\n",
    "When it is finished, you have a file called `Fol4287__vs__Fg_PH1.promer_def.delta` in the directory `results/genome_alignments/`. Open this file and have a look at it's contents.  \n",
    "\n",
    "It should look something like this:  \n",
    "`PROMER  \n",
    "\\>CM000591.1 CM000574.1 5634991 11697295  \n",
    "173055 174734 11476307 11474640 200 129 2  \n",
    "412  \n",
    "1  \n",
    "1  \n",
    "1  \n",
    "0  \n",
    "5329501 5328974 2528 3052 65 45 14  `\n",
    "  \n",
    "We won't go further into any detail, but notice that this is not an intuitive format to read. Luckily MUMmer also has a more manageable format, and we can obtain this by running `show-coords`. We add the `-r` option, to sort the output by reference:  \n",
    "\n",
    "`path/to/MUMmer/show-coords -r /results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def.delta > results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def.coords`\n",
    "  \n",
    ">We will visualize this alignment using Circos, but MUMmer also has a method to quickly display alignments, called MUMmerplot. This needs gnuplot, a plotting program that you can install with `conda`. We won't go into further detail for now but I want show you what this plot looks like:  \n",
    "> <p align=\"center\">\n",
    "<img src=\"out.png\" width=\"600px\" >\n",
    "</p>\n",
    "  \n",
    ">NB I modified the .gp a little to get this exact plot, I made the points a little smaller and removed the mouse-over options as they raised an error...  \n",
    "  \n",
    ">Take a look at this plot, so you a little what to expect.  You can see for example that _Fg_ is on the y-axis (this matches with how we called `nucmer`, we first put the fasta of _Fo_ and then of _Fg_: _Fo_ is considered the `reference` (`REF` on the x-axis) and _Fg_ the `query` (`QRY` on the y-axis).  You can see that chromosome 1 (CM000574.1) in _Fg_ corresponds to chromosome 1 (CM000589.1) and chromosome 7 (CM000595.1) in _Fo_.  We can also see that _Fo_ has a few chromosomes that do not resemble any chromsome in _Fg_: chromosome 3, (CM000591.1) 6 (CM000594.1), 14 (CM000602.2) and 15 (CM000603.1) have very few dots and no lines in this plot.  \n",
    "\n",
    "Now let's visualize this in Circos!  \n",
    "\n",
    "### 3. Use Circos to visualize homologous regions\n",
    "As you may have noticed, _F. graminearum_ has 4 chromosomes where _F. oxysporum_ has 15. This means that chromosomes have rearranged, some may merged in _Fg_, some may have split in _Fo_ or more complicated scenario's of breaking, merging and mixing may have occurred. We already get a nice idea from the dotplot of what is going on, let's visualzie in Circos for perhaps a more clear and more attractive view.\n",
    "\n",
    "Circos uses a config file (ends with `.conf`) that contains information on how a plot should look. First thing we should add to the .conf file is the karyotype: the number and size of chromosomes. This is stored in a karyotype file.\n",
    "\n",
    "### 3.1 Karyotype file\n",
    "The karyotype file also contain information on the colors of the chromosomes:  \n",
    ">chr - ID LABEL START END COLOR\n",
    "\n",
    "So for example, chr 1 in _Fg_ is `CM000574.1`, and it has 11697295 basepairs and if we want plot these in orange we write:  \n",
    "`chr - CM000574.1 1 0 11697295 orange`  \n",
    "  \n",
    "Let's create a such a file using Python. We will plot the chromosomes of different species in different colors in order to separate them. \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fol4287 green\n",
      "CM000589.1 1\n",
      "CM000590.1 2\n",
      "CM000591.1 3\n",
      "CM000592.1 4\n",
      "CM000593.1 5\n",
      "CM000594.1 6\n",
      "CM000595.1 7\n",
      "CM000596.1 8\n",
      "CM000597.1 9\n",
      "CM000598.1 10\n",
      "CM000599.1 11\n",
      "CM000600.1 12\n",
      "CM000601.1 13\n",
      "CM000602.2 14\n",
      "CM000603.1 15\n",
      "Fg_PH1 orange\n",
      "CM000574.1 1\n",
      "CM000575.1 2\n",
      "CM000576.1 3\n",
      "CM000577.1 4\n"
     ]
    }
   ],
   "source": [
    "# first karyotype file\n",
    "import glob\n",
    "\n",
    "# I run this notebook from the working directory hence I don't need to\n",
    "# specify that path; instead of path/to/CircosPlot_FgFo/data/genomes/\n",
    "# I can just say data/genomes/\n",
    "indir   = 'data/genomes/' \n",
    "outdir  = 'circos/input/karyotypes/'\n",
    "def karyotype_file_from_fasta(fasta_fname, out_fname, color = 'grey'):\n",
    "    '''\n",
    "    This function takes a fastafile (fasta_fname) and \"translates\" this \n",
    "    to a karyotype file (out_fname). Grey is the default color for chromosomes.\n",
    "    '''\n",
    "    outfile = open(out_fname, 'w')\n",
    "    l      = 0 # counter for the sequence length\n",
    "    cindex = 1 # counter for the chromosomes, note that this only works \n",
    "               # because the fastafile is ordered!\n",
    "    cid    = ''\n",
    "    for line in open(fasta_fname).readlines():\n",
    "        if line[0] == '>': \n",
    "            '''\n",
    "            This is the fastaheader. We need to store the name in this header \n",
    "            (we assign it to the variable cid), while we count the length of \n",
    "            the subsequent lines (the sequence of this chromosome), until we \n",
    "            encounter the next header\n",
    "            '''\n",
    "            if l > 0: \n",
    "                '''\n",
    "                If l > 0 it means it is not the first sequence.\n",
    "                We first have to process the previous sequence before we can move on \n",
    "                with this one\n",
    "                '''    \n",
    "                #process previous sequence\n",
    "                outfile.write('chr - '+cid+' '+str(cindex)+' 0 '+' '+str(l)+' '+color+'\\n')\n",
    "                #make everything ready for this one\n",
    "                l = 0\n",
    "                cindex += 1\n",
    "                \n",
    "            '''If you encounter a header, store it in cid.\n",
    "            '''\n",
    "            cid = line[1:].strip().split()[0]\n",
    "            print(cid, cindex)\n",
    "                \n",
    "        else:\n",
    "            ''' \n",
    "            If it is not header it is part of the sequence, add the length of this line. \n",
    "            str.strip() takes of any newlines, spaces etc. so we don't count those.\n",
    "            '''\n",
    "            l += len(line.strip())\n",
    "        \n",
    "    ''' \n",
    "    When you've reached the end of the file, you still need to process the last header. \n",
    "    We do this here.\n",
    "    '''\n",
    "    if l > 0:\n",
    "        outfile.write('chr - '+cid+' '+str(cindex)+' 0 '+' '+str(l)+' '+color+'\\n')\n",
    "            \n",
    "    outfile.close()\n",
    "    \n",
    "colors = ['green', 'orange']\n",
    "filenames = glob.glob(indir+'*.fasta')\n",
    "\n",
    "if len(colors) != len(filenames): #check if your list of colors is long enough \n",
    "    print(\"I need more colors, I have \", len(colors), \" color(s) for \", len(filenames), \" files\")\n",
    "else:  \n",
    "    for i, fasta_fname in enumerate(glob.glob(indir+'*.fasta')):\n",
    "        print(fasta_fname.split('/')[-1].split('.fasta')[0],colors[i])\n",
    "        out_fname = outdir + fasta_fname.split('/')[-1].split('.fasta')[0]+'.txt'\n",
    "        karyotype_file_from_fasta(fasta_fname, out_fname, color = colors[i])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Config file\n",
    "#### 3.2.1 Karyotypes and labels\n",
    "Let's see if we can plot these chromosomes in Circos. Circos has so many possibilities that we can not simply define all our wishes on the commandline. They're to many. Circos stores parameters settings in config files. There is one main config file, that can import/load/inlcude other config files. \n",
    "The minimum config file can be found [here](http://circos.ca/documentation/tutorials/quick_start/hello_world/).  (I added line numbers for convenience).  \n",
    "\n",
    "    1:  # circos.conf\n",
    "    2:  \n",
    "    3:  karyotype = data/karyotype/karyotype.human.txt\n",
    "    4:  \n",
    "    5:  <ideogram>\n",
    "    6:  \n",
    "    7:  <spacing\\>\n",
    "    8:  default = 0.005r\n",
    "    9:  </spacing\\>\n",
    "    10:  \n",
    "    11: radius    = 0.9r  \n",
    "    12: thickness = 20p  \n",
    "    13: fill      = yes  \n",
    "    14:  \n",
    "    15: <ideogram\\>\n",
    "    16:    \n",
    "    17: ################################################################    \n",
    "    18: # The remaining content is standard and required. It is imported   \n",
    "    19: # from default files in the Circos distribution.  \n",
    "    20: #  \n",
    "    21: # These should be present in every Circos configuration file and  \n",
    "    22: # overridden as required. To see the content of these files,   \n",
    "    23: # look in etc/ in the Circos distribution.  \n",
    "    24:   \n",
    "    25: <image\\>  \n",
    "    26: # Included from Circos distribution.  \n",
    "    27: <include etc/image.conf>>  \n",
    "    28: </image>  \n",
    "    29:   \n",
    "    30: # RGB/HSV color definitions, color lists, location of fonts, fill patterns.  \n",
    "    31: # Included from Circos distribution.  \n",
    "    32: <<include etc/colors_fonts_patterns.conf>>  \n",
    "    33:     \n",
    "    34: # Debugging, I/O an dother system parameters  \n",
    "    35: # Included from Circos distribution.  \n",
    "    36: <<include etc/housekeeping.conf>>  \n",
    " \n",
    "  \n",
    "  \n",
    "Open a text editor and copy-paste the contents of a minimum config file. Save it with a sensible filename such as Fo_vs_Fg.circos.conf.   \n",
    "\n",
    "First we change the first line, which contains the name of this file.\n",
    "Then we add the correct filename for the karyotype file (line 3). You can plot multiple karyotype files into a single plot by separating them with a ,.\n",
    "\n",
    "You can use radius (line 11) and thickness (line 12) to lot the chromosome in smaller circle or with thicker bands respectively. We also want to add labels to our chromosomes, so we know what we look at.\n",
    "\n",
    "To add labels, you have to add `show_labels = yes` to the configfile and you need to specify where (`label_radius   = dims(ideogram,radius_outer) + 10p`) and how (`label_font = default`, `label_size = 24p`, `label_parallel = yes`) to show the labels. \n",
    "\n",
    "After this, the file becomes:\n",
    "\n",
    "    1:  # Fo_vs_Fg.circos.conf.\n",
    "    2:  \n",
    "    3:  karyotype = circos/input/karyotypes/Fg_PH1.karyotype.txt, circos/input/karyotypes/Fol4287.karyotype.txt\n",
    "    4:  \n",
    "    5:  <ideogram>\n",
    "    6:  \n",
    "    7:  <spacing\\>\n",
    "    8:  default = 0.005r\n",
    "    9:  </spacing\\>\n",
    "    10:  \n",
    "    11: radius    = 0.9r  \n",
    "    12: thickness = 20p  \n",
    "    13: fill      = yes  \n",
    "    14: \n",
    "    15: show_label     = yes\n",
    "    16: label_radius   = dims(ideogram,radius_outer) + 10p\n",
    "    17: label_font     = default\n",
    "    18: label_size     = 24p\n",
    "    19: label_parallel = yes\n",
    "    20: <ideogram\\>\n",
    "    21:    \n",
    "    22: ################################################################    \n",
    "    23: # The remaining content is standard and required. It is imported   \n",
    "    24: # from default files in the Circos distribution.  \n",
    "    25: #  \n",
    "    26: # These should be present in every Circos configuration file and  \n",
    "    27: # overridden as required. To see the content of these files,   \n",
    "    28: # look in etc/ in the Circos distribution.  \n",
    "    29:   \n",
    "    30: <image\\>  \n",
    "    31: # Included from Circos distribution.  \n",
    "    32: <include etc/image.conf>>  \n",
    "    33: </image>  \n",
    "    34:   \n",
    "    35: # RGB/HSV color definitions, color lists, location of fonts, fill patterns.  \n",
    "    36: # Included from Circos distribution.  \n",
    "    37: <<include etc/colors_fonts_patterns.conf>>  \n",
    "    38:     \n",
    "    39: # Debugging, I/O an dother system parameters  \n",
    "    40: # Included from Circos distribution.  \n",
    "    41: <<include etc/housekeeping.conf>>  \n",
    "\n",
    "Let's try!   \n",
    "\n",
    "We tell Circos to use this config file with `-conf your/configfile`.\n",
    "\n",
    "I have installed circos is the `/Applications` folder (MacOS), change this to your Circos installation folder.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "debuggroup summary 0.17s welcome to circos v0.69-6 31 July 2017 on Perl 5.018002\n",
      "debuggroup summary 0.18s current working directory /Users/like/Dropbox/00.Projects/CircosPlot_FgFo\n",
      "debuggroup summary 0.18s command /Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.conf\n",
      "debuggroup summary 0.18s loading configuration from file circos/input/config_files/Fo_vs_Fg.circos.conf\n",
      "debuggroup summary 0.18s found conf file circos/input/config_files/Fo_vs_Fg.circos.conf\n",
      "debuggroup summary 0.33s debug will appear for these features: output,summary\n",
      "debuggroup summary 0.33s bitmap output image ./circos.png\n",
      "debuggroup summary 0.33s SVG output image ./circos.svg\n",
      "debuggroup summary 0.33s parsing karyotype and organizing ideograms\n",
      "debuggroup summary 0.34s karyotype has 19 chromosomes of total size 94,271,448\n",
      "debuggroup summary 0.34s applying global and local scaling\n",
      "debuggroup summary 0.36s allocating image, colors and brushes\n",
      "debuggroup summary 7.47s drawing 19 ideograms of total size 94,271,448\n",
      "debuggroup summary 7.47s drawing highlights and ideograms\n",
      "debuggroup output 7.53s generating output\n",
      "debuggroup output 7.85s created PNG image ./circos.png (106 kb)\n",
      "debuggroup output 7.85s created SVG image ./circos.svg (9 kb)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.conf \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Everything seemed to have worked well, now where is the plot?\n",
    "From the last lines of the output:  \n",
    "\n",
    "    debuggroup output 7.53s created PNG image ./circos.png (106 kb)  \n",
    "    debuggroup output 7.53s created SVG image ./circos.svg (9 kb)  \n",
    "   \n",
    "We can see that Circos saved the figures in the folder it is run from. \n",
    "\n",
    "Let's have a look: \n",
    "<p align=\"center\">\n",
    "<img src=\"circos_karyotype_with_labels.png\" width=\"600px\" >\n",
    "</p>\n",
    "\n",
    "\n",
    "We can change this either in the configfile (see an other notebook Circos_output_config_files on how to do this and understand better how Circos uses config files) or on the commandline. Let's do that:  \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "debuggroup summary 0.24s welcome to circos v0.69-6 31 July 2017 on Perl 5.018002\n",
      "debuggroup summary 0.24s current working directory /Users/like/Dropbox/00.Projects/CircosPlot_FgFo\n",
      "debuggroup summary 0.24s command /Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.conf -outputdir circos/output -outputfile circos_karyotype_with_labels.png\n",
      "debuggroup summary 0.24s loading configuration from file circos/input/config_files/Fo_vs_Fg.circos.conf\n",
      "debuggroup summary 0.24s found conf file circos/input/config_files/Fo_vs_Fg.circos.conf\n",
      "debuggroup summary 0.39s debug will appear for these features: output,summary\n",
      "debuggroup summary 0.39s bitmap output image circos/output/circos_karyotype_with_labels.png\n",
      "debuggroup summary 0.39s SVG output image circos/output/circos_karyotype_with_labels.svg\n",
      "debuggroup summary 0.39s parsing karyotype and organizing ideograms\n",
      "debuggroup summary 0.39s karyotype has 19 chromosomes of total size 94,271,448\n",
      "debuggroup summary 0.40s applying global and local scaling\n",
      "debuggroup summary 0.41s allocating image, colors and brushes\n",
      "debuggroup summary 2.39s drawing 19 ideograms of total size 94,271,448\n",
      "debuggroup summary 2.39s drawing highlights and ideograms\n",
      "debuggroup output 2.43s generating output\n",
      "debuggroup output 2.76s created PNG image circos/output/circos_karyotype_with_labels.png (106 kb)\n",
      "debuggroup output 2.76s created SVG image circos/output/circos_karyotype_with_labels.svg (9 kb)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.conf -outputdir circos/output -outputfile circos_karyotype_with_labels.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Links\n",
    "#### 3.3.1 Write links file from MUMmer coords file\n",
    "We now have plotted the chromosomes of _Fo_ and _Fg_, but we want to see whther and how different chromosomes in these two species correspond to eachother. Circos is desiged to show links between ordered sets, so let's tart using it what is was designed to do.  \n",
    "  \n",
    "Link files have the [following format](http://circos.ca/documentation/tutorials/links/basic_links/lesson):  \n",
    ">chrA from to chrB from to\n",
    "\n",
    "We can get this information from the coords file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    9030     9248  |  7107985  7108206  |      219      222  |    63.51    68.92     0.68  |  3  1  CM000589.1\tCM000574.1\n",
      "\n",
      "['9030', '9248', '|', '7107985', '7108206', '|', '219', '222', '|', '63.51', '68.92', '0.68', '|', '3', '1', 'CM000589.1', 'CM000574.1']\n"
     ]
    }
   ],
   "source": [
    "lines   = open('results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def.coords').readlines()[5:]\n",
    "print(lines[0])\n",
    "columns = lines[0].split()\n",
    "print(columns)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's write the 'translation' from this format to the circos link format to a function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def.coords : Done\n"
     ]
    }
   ],
   "source": [
    "def coords_to_links(coords_fname, links_fname):\n",
    "    linksfile = open(links_fname, 'w')\n",
    "    lines     = open(coords_fname).readlines()[5:] #skip header\n",
    "    for line in lines:\n",
    "        columns = line.strip().split()\n",
    "        \n",
    "        if len(columns) != 17:\n",
    "            print('Unexpected fileformat, can not continue!')\n",
    "            break\n",
    "            \n",
    "        chrA  = columns[15]\n",
    "        chrB  = columns[16]\n",
    "        fromA = columns[0]\n",
    "        toA   = columns[1]\n",
    "        fromB = columns[3]\n",
    "        toB   = columns[4]\n",
    "        \n",
    "        linksfile.write(chrA+' '+fromA+' '+toA+' '+chrB+' '+fromB+' '+toB+'\\n')\n",
    "    linksfile.close()\n",
    "    print(coords_fname,\": Done\")\n",
    "    \n",
    "coords_fname ='results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def.coords'\n",
    "links_fname  = 'circos/input/links/links_Fo_vs_Fg.txt'\n",
    "coords_to_links(coords_fname, links_fname)\n",
    "               "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's have a look:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "less circos/input/links/links_Fo_vs_Fg.txt | head -10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3.3.2 Include links in config file\n",
    "Let's add this to our conf file (save it under a new name):\n",
    "\n",
    "    # Fo_vs_Fg.circos.incl_links.conf\n",
    "    \n",
    "    karyotype = karyotypes/Fg_PH1.txt,karyotypes/Fol4287.txt\n",
    "\n",
    "    <ideogram>\n",
    "    <spacing>\n",
    "    \n",
    "    default = 0.005r\n",
    "\n",
    "    </spacing>\n",
    "\n",
    "    #position\n",
    "    radius    \t= 0.80r\n",
    "    thickness \t= 50p\n",
    "    fill     \t= yes\n",
    "\n",
    "    #label\n",
    "    show_label     = yes\n",
    "    label_font     = default\n",
    "    label_radius   = dims(ideogram,radius_outer) + 10p\n",
    "    label_size     = 24p\n",
    "    label_parallel = yes\n",
    "\n",
    "\n",
    "    </ideogram>\n",
    "    \n",
    "**Here we can add a `links` section:**\n",
    "    \n",
    "    <links>\n",
    "\n",
    "    <link>\n",
    "    file          = input/links/links_Fo_vs_Fg.txt\n",
    "    radius        = 0.95r\n",
    "    color         = black_a4\n",
    "\n",
    "    # Curves look best when this value is small (e.g. 0.1r or 0r)\n",
    "    bezier_radius = 0.1r\n",
    "    thickness     = 1\n",
    "\n",
    "    </link>\n",
    "\n",
    "    </links>\n",
    "\n",
    "\n",
    "    ################################################################\n",
    "    # The remaining content is standard and required. It is imported \n",
    "    # from default files in the Circos distribution.\n",
    "    #\n",
    "    # These should be present in every Circos configuration file and\n",
    "    # overridden as required. To see the content of these files, \n",
    "    # look in etc/ in the Circos distribution.\n",
    "\n",
    "    <image>\n",
    "    # Included from Circos distribution.\n",
    "\n",
    "    <<include etc/image.conf>>\n",
    "    </image>\n",
    "\n",
    "\n",
    "    # RGB/HSV color definitions, color lists, location of fonts, fill patterns.\n",
    "    # Included from Circos distribution.\n",
    "    <<include etc/colors_fonts_patterns.conf>>\n",
    "\n",
    "    # Debugging, I/O an dother system parameters\n",
    "    # Included from Circos distribution.\n",
    "    <<include etc/housekeeping.conf>>\n",
    "    \n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "debuggroup summary 0.14s welcome to circos v0.69-6 31 July 2017 on Perl 5.018002\n",
      "debuggroup summary 0.14s current working directory /Users/like/Dropbox/00.Projects/CircosPlot_FgFo\n",
      "debuggroup summary 0.14s command /Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.incl_links.conf -outputdir circos/output -outputfile circos_karyotype_with_labels_and_links.png\n",
      "debuggroup summary 0.14s loading configuration from file circos/input/config_files/Fo_vs_Fg.circos.incl_links.conf\n",
      "debuggroup summary 0.14s found conf file circos/input/config_files/Fo_vs_Fg.circos.incl_links.conf\n",
      "debuggroup summary 0.37s debug will appear for these features: output,summary\n",
      "debuggroup summary 0.37s bitmap output image circos/output/circos_karyotype_with_labels_and_links.png\n",
      "debuggroup summary 0.37s SVG output image circos/output/circos_karyotype_with_labels_and_links.svg\n",
      "debuggroup summary 0.38s parsing karyotype and organizing ideograms\n",
      "debuggroup summary 0.38s karyotype has 19 chromosomes of total size 94,271,448\n",
      "debuggroup summary 0.39s applying global and local scaling\n",
      "debuggroup summary 0.40s allocating image, colors and brushes\n",
      "debuggroup summary 2.62s drawing 19 ideograms of total size 94,271,448\n",
      "debuggroup summary 2.62s drawing highlights and ideograms\n",
      "debuggroup summary 2.68s found conf file /Applications/circos-0.69-6/bin/../etc/tracks/link.conf\n",
      "debuggroup summary 2.68s process track_0 link circos/input/config_files/../../input/links/links_Fo_vs_Fg.txt\n",
      "debuggroup summary 11.09s drawing link track_0 z 0\n",
      "\n",
      "  *** CIRCOS ERROR ***\n",
      "\n",
      "      cwd: /Users/like/Dropbox/00.Projects/CircosPlot_FgFo\n",
      "\n",
      "      command: /Applications/circos-0.69-6/bin/circos -conf\n",
      "      circos/input/config_files/Fo_vs_Fg.circos.incl_links.conf -outputdir\n",
      "      circos/output -outputfile circos_karyotype_with_labels_and_links.png\n",
      "\n",
      "  You have asked to draw [54762] links from file\n",
      "\n",
      "      input/links/links_Fo_vs_Fg.txt\n",
      "\n",
      "  but the maximum is currently set at [25000]. To increase this number change\n",
      "  max_links in etc/housekeeping.conf. Keep in mind that drawing that many links\n",
      "  may create an image that is too busy and uninterpretable.\n",
      "\n",
      "  If you are having trouble debugging this error, first read the best practices\n",
      "  tutorial for helpful tips that address many common problems\n",
      "\n",
      "      http://www.circos.ca/documentation/tutorials/reference/best_practices\n",
      "\n",
      "  The debugging facility is helpful to figure out what's happening under the\n",
      "  hood\n",
      "\n",
      "      http://www.circos.ca/documentation/tutorials/configuration/debugging\n",
      "\n",
      "  If you're still stumped, get support in the Circos Google Group.\n",
      "\n",
      "      http://groups.google.com/group/circos-data-visualization\n",
      "\n",
      "  Please include this error, all your configuration, data files and the version\n",
      "  of Circos you're running (circos -v).Do not email me directly -- please use\n",
      "  the group.\n",
      "\n",
      "  Stack trace:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " at /Applications/circos-0.69-6/bin/../lib/Circos/Error.pm line 425, <F> line 54762.\n",
      "\tCircos::Error::fatal_error('links', 'max_number', 54762, 'input/links/links_Fo_vs_Fg.txt', 25000) called at /Applications/circos-0.69-6/bin/../lib/Circos.pm line 1210\n",
      "\tCircos::run('Circos', 'outputdir', 'circos/output', 'outputfile', 'circos_karyotype_with_labels_and_links.png', '_argv', '-conf circos/input/config_files/Fo_vs_Fg.circos.incl_links.co...', '_cwd', '/Users/like/Dropbox/00.Projects/CircosPlot_FgFo', ...) called at /Applications/circos-0.69-6/bin/circos line 536\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.incl_links.conf  -outputdir circos/output -outputfile circos_karyotype_with_labels_and_links.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "We get an error that we're trying to display to many links. We can either reset the maximum number of links, or we can filter the links we plot. Let's do the latter:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def coords_to_links(coords_fname, links_fname, min_length = 0):\n",
    "    linksfile = open(links_fname, 'w')\n",
    "    lines     = open(coords_fname).readlines()[5:] #skip header\n",
    "    for line in lines:\n",
    "        columns = line.strip().split()\n",
    "        \n",
    "        if len(columns) != 17:\n",
    "            print('Unexpected fileformat, can not continue!')\n",
    "            break\n",
    "            \n",
    "        chrA  = columns[15]\n",
    "        chrB  = columns[16]\n",
    "        fromA = columns[0]\n",
    "        toA   = columns[1]\n",
    "        fromB = columns[3]\n",
    "        toB   = columns[4]\n",
    "        \n",
    "        lenA  = int(columns[6])\n",
    "        lenB  = int(columns[7])\n",
    "        if lenA >= min_length and lenB >= min_length: \n",
    "            linksfile.write(chrA+' '+fromA+' '+toA+' '+chrB+' '+fromB+' '+toB+'\\n')\n",
    "    linksfile.close()\n",
    "\n",
    "coords_fname = 'results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def.coords'\n",
    "min_length   = 1000\n",
    "links_fname  = 'circos/input/links/links_Fo_vs_Fg.min_length'+str(min_length)+'bp.txt'\n",
    "coords_to_links(coords_fname, links_fname, min_length = min_length)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we filter links based on their length, only including alignments that span 1000 bp or more in both genomes. ALternatively, you can also filter based on sequence similarity or both.\n",
    "\n",
    "Let's change our config file to display the filtered links only (again saving the file under a new name):\n",
    "\n",
    "    # Fo_vs_Fg.circos.incl_links.filtered.conf\n",
    "    \n",
    "    karyotype = input/karyotypes/Fg_PH1.txt, input/karyotypes/Fol4287.txt\n",
    "\n",
    "    <ideogram>\n",
    "    <spacing>\n",
    "    \n",
    "    default = 0.005r\n",
    "\n",
    "    </spacing>\n",
    "\n",
    "    #position\n",
    "    radius    \t= 0.80r\n",
    "    thickness \t= 50p\n",
    "    fill     \t= yes\n",
    "\n",
    "    #label\n",
    "    show_label     = yes\n",
    "    label_font     = default\n",
    "    label_radius   = dims(ideogram,radius_outer) + 10p\n",
    "    label_size     = 24p\n",
    "    label_parallel = yes\n",
    "\n",
    "\n",
    "    </ideogram>\n",
    "    \n",
    "\n",
    "    \n",
    "    <links>\n",
    "\n",
    "    <link>\n",
    "**Change the file in the `links` section:**\n",
    "\n",
    "    file          = input/links/links_Fo_vs_Fg.min_length1000bp.txt\n",
    "    radius        = 0.95r\n",
    "    color         = black_a4\n",
    "\n",
    "    # Curves look best when this value is small (e.g. 0.1r or 0r)\n",
    "    bezier_radius = 0.1r\n",
    "    thickness     = 1\n",
    "\n",
    "    </link>\n",
    "\n",
    "    </links>\n",
    "\n",
    "\n",
    "    ################################################################\n",
    "    # The remaining content is standard and required. It is imported \n",
    "    # from default files in the Circos distribution.\n",
    "    #\n",
    "    # These should be present in every Circos configuration file and\n",
    "    # overridden as required. To see the content of these files, \n",
    "    # look in etc/ in the Circos distribution.\n",
    "\n",
    "    <image>\n",
    "    # Included from Circos distribution.\n",
    "    #dir  = /path/to/your/output/directory\n",
    "    #file = yourimage.png\n",
    "\n",
    "    <<include etc/image.conf>>\n",
    "    </image>\n",
    "\n",
    "\n",
    "    # RGB/HSV color definitions, color lists, location of fonts, fill patterns.\n",
    "    # Included from Circos distribution.\n",
    "    <<include etc/colors_fonts_patterns.conf>>\n",
    "\n",
    "    # Debugging, I/O an dother system parameters\n",
    "    # Included from Circos distribution.\n",
    "    <<include etc/housekeeping.conf>>\n",
    "    \n",
    "\n",
    "Let's try and run this (this may take a while, be a little patient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "debuggroup summary 0.15s welcome to circos v0.69-6 31 July 2017 on Perl 5.018002\n",
      "debuggroup summary 0.15s current working directory /Users/like/Dropbox/00.Projects/CircosPlot_FgFo\n",
      "debuggroup summary 0.15s command /Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.incl_links.filtered.conf -outputdir circos/output -outputfile circos_karyotype_with_labels_and_links.png\n",
      "debuggroup summary 0.15s loading configuration from file circos/input/config_files/Fo_vs_Fg.circos.incl_links.filtered.conf\n",
      "debuggroup summary 0.15s found conf file circos/input/config_files/Fo_vs_Fg.circos.incl_links.filtered.conf\n",
      "debuggroup summary 0.31s debug will appear for these features: output,summary\n",
      "debuggroup summary 0.32s bitmap output image circos/output/circos_karyotype_with_labels_and_links.png\n",
      "debuggroup summary 0.32s SVG output image circos/output/circos_karyotype_with_labels_and_links.svg\n",
      "debuggroup summary 0.32s parsing karyotype and organizing ideograms\n",
      "debuggroup summary 0.32s karyotype has 19 chromosomes of total size 94,271,448\n",
      "debuggroup summary 0.32s applying global and local scaling\n",
      "debuggroup summary 0.33s allocating image, colors and brushes\n",
      "debuggroup summary 2.13s drawing 19 ideograms of total size 94,271,448\n",
      "debuggroup summary 2.13s drawing highlights and ideograms\n",
      "debuggroup summary 2.18s found conf file /Applications/circos-0.69-6/bin/../etc/tracks/link.conf\n",
      "debuggroup summary 2.18s process track_0 link circos/input/config_files/../../input/links/links_Fo_vs_Fg.min_length1000bp.txt\n",
      "debuggroup summary 4.50s drawing link track_0 z 0\n",
      "debuggroup output 20.42s generating output\n",
      "debuggroup output 21.01s created PNG image circos/output/circos_karyotype_with_labels_and_links.png (2041 kb)\n",
      "debuggroup output 21.01s created SVG image circos/output/circos_karyotype_with_labels_and_links.svg (11311 kb)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.incl_links.filtered.conf  -outputdir circos/output -outputfile circos_karyotype_with_labels_and_links.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p align=\"center\">\n",
    "<img src=\"circos_karyotype_with_labels_and_links.png\" width=\"600px\" >\n",
    "</p>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CM000589.1 1\n",
      "CM000590.1 2\n",
      "CM000591.1 3\n",
      "CM000592.1 4\n",
      "CM000593.1 5\n",
      "CM000594.1 6\n",
      "CM000595.1 7\n",
      "CM000596.1 8\n",
      "CM000597.1 9\n",
      "CM000598.1 10\n",
      "CM000599.1 11\n",
      "CM000600.1 12\n",
      "CM000601.1 13\n",
      "CM000602.2 14\n",
      "CM000603.1 15\n",
      "CM000574.1 1\n",
      "CM000575.1 2\n",
      "CM000576.1 3\n",
      "CM000577.1 4\n"
     ]
    }
   ],
   "source": [
    "import glob\n",
    "\n",
    "def coords_to_links(coords_fname, links_fname, chr2color = None, min_length = 0):\n",
    "    linksfile = open(links_fname, 'w')\n",
    "    lines     = open(coords_fname).readlines()[5:] #skip header\n",
    "    for line in lines:\n",
    "        columns = line.strip().split()\n",
    "        \n",
    "        if len(columns) != 17:\n",
    "            print('Unexpected fileformat, can not continue!')\n",
    "            break\n",
    "            \n",
    "        chrA  = columns[15]\n",
    "        chrB  = columns[16]\n",
    "        fromA = columns[0]\n",
    "        toA   = columns[1]\n",
    "        fromB = columns[3]\n",
    "        toB   = columns[4]\n",
    "        \n",
    "        lenA  = int(columns[6])\n",
    "        lenB  = int(columns[7])\n",
    "        if lenA >= min_length and lenB >= min_length: \n",
    "            out = chrA+' '+fromA+' '+toA+' '+chrB+' '+fromB+' '+toB\n",
    "            \n",
    "            if chr2color != None:\n",
    "                if chrA in chr2color:\n",
    "                    linksfile.write(out+' color='+chr2color[chrA]+'\\n')\n",
    "                elif chrB in chr2color:\n",
    "                    linksfile.write(out+' color='+chr2color[chrB]+'\\n')\n",
    "            else:\n",
    "                linksfile.write(out+'\\n')\n",
    "    linksfile.close()\n",
    "\n",
    "\n",
    "def karyotype_file_from_fasta(fasta_fname, out_fname, chr2color = None):\n",
    "\n",
    "    outfile = open(out_fname, 'w')\n",
    "    l      = 0 #counter for the sequence length\n",
    "    cindex = 1 #counter for the chromosomes\n",
    "    cid    = ''\n",
    "    color  = 'grey'\n",
    "    for line in open(fasta_fname).readlines():\n",
    "        if line[0] == '>':\n",
    "            if l > 0:\n",
    "                outfile.write('chr - '+cid+' '+str(cindex)+' 0 '+' '+str(l)+' '+color+'\\n')\n",
    "                l = 0\n",
    "                cindex += 1\n",
    "            cid = line[1:].strip().split()[0]\n",
    "            if chr2color != None and cid in chr2color:\n",
    "                color = chr2color[cid]\n",
    "            print(cid, cindex)\n",
    "                \n",
    "        else:\n",
    "            l += len(line.strip())\n",
    "        \n",
    "    #last one:\n",
    "    if l > 0:\n",
    "        outfile.write('chr - '+cid+' '+str(cindex)+' 0 '+' '+str(l)+' '+color+'\\n')\n",
    "            \n",
    "    outfile.close()\n",
    "    \n",
    "# write karyotype files from fasta\n",
    "chr2color = {}\n",
    "chr2color['CM000574.1'] = 'blue'\n",
    "chr2color['CM000575.1'] = 'yellow'\n",
    "chr2color['CM000576.1'] = 'green'\n",
    "chr2color['CM000577.1'] = 'red'\n",
    "\n",
    "for i, fasta_fname in enumerate(glob.glob('data/genomes/*.fasta')):\n",
    "    \n",
    "    out_fname = 'circos/input/karyotypes/' + fasta_fname.split('/')[-1].split('.fasta')[0]+'.color-coded.txt'\n",
    "    karyotype_file_from_fasta(fasta_fname, out_fname, chr2color = chr2color)\n",
    "\n",
    "coords_fname = 'results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def.coords'\n",
    "links_fname  = 'circos/input/links/links_Fo_vs_Fg.min_length1000bp.color-coded.txt'\n",
    "coords_to_links(coords_fname, links_fname, chr2color = chr2color, min_length = 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now adjust the config file, to incorporate these karyotype and links files:\n",
    "    \n",
    "    # Fo_vs_Fg.circos.incl_colored_links.filtered.conf\n",
    "    \n",
    "    karyotype = input/karyotypes/Fg_PH1.color-coded.txt, input/karyotypes/Fol4287.color-coded.txt\n",
    "\n",
    "    <ideogram>\n",
    "    <spacing>\n",
    "    \n",
    "    default = 0.005r\n",
    "\n",
    "    </spacing>\n",
    "\n",
    "    #position\n",
    "    radius    \t= 0.80r\n",
    "    thickness \t= 50p\n",
    "    fill     \t= yes\n",
    "\n",
    "    #label\n",
    "    show_label     = yes\n",
    "    label_font     = default\n",
    "    label_radius   = dims(ideogram,radius_outer) + 10p\n",
    "    label_size     = 24p\n",
    "    label_parallel = yes\n",
    "\n",
    "\n",
    "    </ideogram>\n",
    "    \n",
    "\n",
    "    \n",
    "    <links>\n",
    "\n",
    "    <link>\n",
    "**Change the file in the `links` section:**\n",
    "\n",
    "    file          = input/links/links_Fo_vs_Fg.min_length1000bp.color-coded.txt\n",
    "    radius        = 0.95r\n",
    "    color         = black_a4\n",
    "\n",
    "    # Curves look best when this value is small (e.g. 0.1r or 0r)\n",
    "    bezier_radius = 0.1r\n",
    "    thickness     = 1\n",
    "\n",
    "    </link>\n",
    "\n",
    "    </links>\n",
    "\n",
    "\n",
    "    ################################################################\n",
    "    # The remaining content is standard and required. It is imported \n",
    "    # from default files in the Circos distribution.\n",
    "    #\n",
    "    # These should be present in every Circos configuration file and\n",
    "    # overridden as required. To see the content of these files, \n",
    "    # look in etc/ in the Circos distribution.\n",
    "\n",
    "    <image>\n",
    "    # Included from Circos distribution.\n",
    "    #dir  = /path/to/your/output/directory\n",
    "    #file = yourimage.png\n",
    "\n",
    "    <<include etc/image.conf>>\n",
    "    </image>\n",
    "\n",
    "\n",
    "    # RGB/HSV color definitions, color lists, location of fonts, fill patterns.\n",
    "    # Included from Circos distribution.\n",
    "    <<include etc/colors_fonts_patterns.conf>>\n",
    "\n",
    "    # Debugging, I/O an dother system parameters\n",
    "    # Included from Circos distribution.\n",
    "    <<include etc/housekeeping.conf>>\n",
    "    \n",
    "\n",
    "Let's try and run this (this may take a while, be a little patient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "debuggroup summary 0.12s welcome to circos v0.69-6 31 July 2017 on Perl 5.018002\n",
      "debuggroup summary 0.13s current working directory /Users/like/Dropbox/00.Projects/CircosPlot_FgFo\n",
      "debuggroup summary 0.13s command /Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.incl_colored_links.filtered.conf -outputdir circos/output -outputfile circos_karyotype_with_labels_and_colored_links.png\n",
      "debuggroup summary 0.13s loading configuration from file circos/input/config_files/Fo_vs_Fg.circos.incl_colored_links.filtered.conf\n",
      "debuggroup summary 0.13s found conf file circos/input/config_files/Fo_vs_Fg.circos.incl_colored_links.filtered.conf\n",
      "debuggroup summary 0.27s debug will appear for these features: output,summary\n",
      "debuggroup summary 0.28s bitmap output image circos/output/circos_karyotype_with_labels_and_colored_links.png\n",
      "debuggroup summary 0.28s SVG output image circos/output/circos_karyotype_with_labels_and_colored_links.svg\n",
      "debuggroup summary 0.28s parsing karyotype and organizing ideograms\n",
      "debuggroup summary 0.28s karyotype has 19 chromosomes of total size 94,271,448\n",
      "debuggroup summary 0.28s applying global and local scaling\n",
      "debuggroup summary 0.30s allocating image, colors and brushes\n",
      "debuggroup summary 2.26s drawing 19 ideograms of total size 94,271,448\n",
      "debuggroup summary 2.26s drawing highlights and ideograms\n",
      "debuggroup summary 2.30s found conf file /Applications/circos-0.69-6/bin/../etc/tracks/link.conf\n",
      "debuggroup summary 2.30s process track_0 link circos/input/config_files/../../input/links/links_Fo_vs_Fg.min_length1000bp.color-coded.txt\n",
      "debuggroup summary 4.98s drawing link track_0 z 0\n",
      "debuggroup output 20.76s generating output\n",
      "debuggroup output 21.40s created PNG image circos/output/circos_karyotype_with_labels_and_colored_links.png (3771 kb)\n",
      "debuggroup output 21.40s created SVG image circos/output/circos_karyotype_with_labels_and_colored_links.svg (11407 kb)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg.circos.incl_colored_links.filtered.conf  -outputdir circos/output -outputfile circos_karyotype_with_labels_and_colored_links.png\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p align=\"center\">\n",
    "<img src=\"circos_karyotype_with_labels_and_colored_links.png\" width=\"600px\" >\n",
    "</p>\n",
    "\n",
    "This is getting somewhere! We can see that chromosom1 in _Fg_ corresponds roughly to chr 7, half of chr 4 and chr 1 in _Fo_, but e.g. for chr 02 in _Fg_ it is tricky to see which part belongs to which chromosome in _Fo_. We can further improve our visualization by color-coding the chromosomes in _Fo_, rather than the chromosomes in _Fg_ and make 4 plots: one per chromosome.\n",
    "\n",
    "Circos implements [brewer palettes for qualitative data] (http://colorbrewer2.org/#type=qualitative&scheme=Set3&n=12) for up to 12 categories. _Fo_ has 15 chromosomes, but chromosome 3, 6, 14 and 15 have very little regions that are homologous to regions in _Fg_ (are lineage-specific), hence we use a single color (e.g. black) for those. \n",
    "\n",
    "If we use `set3` of 11 colors (15 - 4 lineage-specific chromosomes) corresponding to categories (rather e.g. sequential data):`qual`, we can refer to them with `set3-11-qual-N`, where N is the number 1 to 11. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CM000589.1 1\n",
      "CM000590.1 2\n",
      "CM000591.1 3\n",
      "CM000592.1 4\n",
      "CM000593.1 5\n",
      "CM000594.1 6\n",
      "CM000595.1 7\n",
      "CM000596.1 8\n",
      "CM000597.1 9\n",
      "CM000598.1 10\n",
      "CM000599.1 11\n",
      "CM000600.1 12\n",
      "CM000601.1 13\n",
      "CM000602.2 14\n",
      "CM000603.1 15\n",
      "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg_chr1.circos.incl_colored_links.filtered.conf -outputdir circos/output -outputfile circos_Fg_chr1.vs.Fol4287_with_labels_and_colored_links.png\n",
      "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg_chr2.circos.incl_colored_links.filtered.conf -outputdir circos/output -outputfile circos_Fg_chr2.vs.Fol4287_with_labels_and_colored_links.png\n",
      "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg_chr3.circos.incl_colored_links.filtered.conf -outputdir circos/output -outputfile circos_Fg_chr3.vs.Fol4287_with_labels_and_colored_links.png\n",
      "/Applications/circos-0.69-6/bin/circos -conf circos/input/config_files/Fo_vs_Fg_chr4.circos.incl_colored_links.filtered.conf -outputdir circos/output -outputfile circos_Fg_chr4.vs.Fol4287_with_labels_and_colored_links.png\n"
     ]
    }
   ],
   "source": [
    "import glob, os\n",
    "\n",
    "# Write karyotype files from fasta\n",
    "chr2color = {}\n",
    "chr2color['CM000589.1'] = 'set3-11-qual-1'\n",
    "chr2color['CM000590.1'] = 'set3-11-qual-2'\n",
    "chr2color['CM000591.1'] = 'light-grey' #chr 3: lineage-specific\n",
    "chr2color['CM000592.1'] = 'set3-11-qual-3'\n",
    "chr2color['CM000593.1'] = 'set3-11-qual-4'\n",
    "chr2color['CM000594.1'] = 'light-grey' #chr 6: lineage-specific\n",
    "chr2color['CM000595.1'] = 'set3-11-qual-5'\n",
    "chr2color['CM000596.1'] = 'set3-11-qual-6'\n",
    "chr2color['CM000597.1'] = 'set3-11-qual-7'\n",
    "chr2color['CM000598.1'] = 'set3-11-qual-8'\n",
    "chr2color['CM000599.1'] = 'set3-11-qual-9'\n",
    "chr2color['CM000600.1'] = 'set3-11-qual-10'\n",
    "chr2color['CM000601.1'] = 'set3-11-qual-11'\n",
    "chr2color['CM000602.2'] = 'light-grey' #chr 14: lineage-specific\n",
    "chr2color['CM000603.1'] = 'light-grey' #chr 15: lineage-specific\n",
    "\n",
    "\n",
    "# Write the karyotype file for Fol4287\n",
    "Fol_karyo_fname = 'circos/input/karyotypes/Fol4287_chr_color-coded.txt'\n",
    "fasta_fname     = 'data/genomes/Fol4287.fasta'\n",
    "karyotype_file_from_fasta(fasta_fname, Fol_karyo_fname, chr2color = chr2color)\n",
    "\n",
    "#write the links file, now color them according to the Fol4287 chromosomes instead of the Fg_PH1 chromosomes\n",
    "coords_fname = 'results/genome_alignments/Fol4287__vs__Fg_PH1.promer_def.coords'\n",
    "links_fname  = 'circos/input/links/links_Fo_vs_Fg.min_length1000bp.color-coded_Fol.txt'\n",
    "coords_to_links(coords_fname, links_fname, chr2color = chr2color, min_length = 1000)\n",
    "\n",
    "# We need to make 4 plots, one for each Fg chromosome.\n",
    "# There are a lot of different ways to do this. \n",
    "#\n",
    "# I decided here to write a separate file for each chromosome in F. graminearum, \n",
    "# splitting up the existing karyotype file.\n",
    "#\n",
    "# Moreover, as I am looping over the chromosomes in the karyotypefile, \n",
    "# I also write the config file.\n",
    "#\n",
    "# First open a 'template' config file and store it's lines, so we can write them to \n",
    "# the 4 new chromosome-specific config files:\n",
    "\n",
    "config_lines = open('circos/input/config_files/Fo_vs_Fg.circos.incl_colored_links.filtered.conf').readlines()\n",
    "\n",
    "'''\n",
    "You can run the code below, to print the line number and line of the config file to make sure \n",
    "you're changing the correct line:\n",
    "\n",
    "for i,line in enumerate(config_lines):\n",
    "    print(i,line.strip())\n",
    "    \n",
    "'''\n",
    "\n",
    "# loop over the Fg chromosomes:\n",
    "karyotype_file_Fg = 'circos/input/karyotypes/Fg_PH1.txt'\n",
    "for i, line in enumerate(open(karyotype_file_Fg).readlines()):\n",
    "    \n",
    "    # write the karyotype file\n",
    "    chrom_spec_fname = 'circos/input/karyotypes/Fg_chr'+str(i+1)+'.txt'\n",
    "    Fg_chrom_file = open(chrom_spec_fname, 'w')\n",
    "    Fg_chrom_file.write(line.replace('orange', 'grey')) # change color of the chromosomes \n",
    "                                                        # (one of the Fo chrs is orange)\n",
    "    Fg_chrom_file.close()\n",
    "    \n",
    "    # Make the config file:\n",
    "    config_fname         = 'circos/input/config_files/Fo_vs_Fg_chr'+str(i+1)\\\n",
    "                                +'.circos.incl_colored_links.filtered.conf'\n",
    "    # change the line with the name of the config file\n",
    "    config_lines[0] = '# '+config_fname+'\\n'\n",
    "    # change the line listing the karyotype files, to use the chr-specific karyotype file for Fg\n",
    "    config_lines[2] = 'karyotype = '+chrom_spec_fname+','+Fol_karyo_fname+'\\n'\n",
    "    # change the line with the file to use for the links\n",
    "    config_lines[31] = 'file\\t= '+links_fname+'\\n'\n",
    "    \n",
    "    # Write the config file\n",
    "    Fg_chrom_config_file = open(config_fname, 'w')\n",
    "    for line in config_lines:\n",
    "        Fg_chrom_config_file.write(line)\n",
    "    Fg_chrom_config_file.close()\n",
    "    \n",
    "    # Finally, run circos using this config file:\n",
    "    circos_command = '/Applications/circos-0.69-6/bin/circos' # path to Circos\n",
    "    circos_command +=' -conf '+config_fname                   # the config file we want circos to use\n",
    "    circos_command +=' -outputdir circos/output'              # where to save the output\n",
    "    # filename of the generated plot\n",
    "    circos_command +=' -outputfile circos_Fg_chr'+str(i+1)+'.vs.Fol4287_with_labels_and_colored_links.png' \n",
    "\n",
    "    print(circos_command)\n",
    "    os.system(circos_command)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"circos_Fg_chr1.vs.Fol4287_with_labels_and_colored_links.png\" style=\"float: left; width: 25%; margin-right: 0%;margin-bottom: 0.5em;\">\n",
    "<img src=\"circos_Fg_chr2.vs.Fol4287_with_labels_and_colored_links.png\" style=\"float: left; width: 25%; margin-right: 0%;margin-bottom: 0.5em;\">\n",
    "<img src=\"circos_Fg_chr3.vs.Fol4287_with_labels_and_colored_links.png\" style=\"float: left; width: 25%; margin-right: 0%;margin-bottom: 0.5em;\">\n",
    "<img src=\"circos_Fg_chr4.vs.Fol4287_with_labels_and_colored_links.png\" style=\"float: left; width: 25%; margin-right: 0%;margin-bottom: 0.5em;\">\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "  \n",
    "   \n",
    "## Note on versioning  \n",
    "\n",
    "During this workshop/tutorial, we've been adjusting our code and circos input and outfiles, constantly improving them.\n",
    "This resulted in the following output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     circos/input/config_files : \n",
      "Fo_vs_Fg.circos.conf\n",
      "Fo_vs_Fg.circos.incl_colored_links.filtered.conf\n",
      "Fo_vs_Fg.circos.incl_links.conf\n",
      "Fo_vs_Fg.circos.incl_links.filtered.conf\n",
      "Fo_vs_Fg.circos.specify_output.conf\n",
      "Fo_vs_Fg_chr1.circos.incl_colored_links.filtered.conf\n",
      "Fo_vs_Fg_chr2.circos.incl_colored_links.filtered.conf\n",
      "Fo_vs_Fg_chr3.circos.incl_colored_links.filtered.conf\n",
      "Fo_vs_Fg_chr4.circos.incl_colored_links.filtered.conf\n",
      "image.Fo_vs_Fg.conf\n",
      "image.generic.Fo_vs_Fg.conf\n",
      "     circos/input/karyotypes : \n",
      "Fg_PH1.color-coded.txt\n",
      "Fg_PH1.txt\n",
      "Fg_chr0.txt\n",
      "Fg_chr1.txt\n",
      "Fg_chr2.txt\n",
      "Fg_chr3.txt\n",
      "Fg_chr4.txt\n",
      "Fol4287.color-coded.txt\n",
      "Fol4287.txt\n",
      "Fol4287_chr_color-coded.txt\n",
      "     circos/input/links : \n",
      "links_Fo_vs_Fg.min_length1000bp.color-coded.txt\n",
      "links_Fo_vs_Fg.min_length1000bp.color-coded_Fol.txt\n",
      "links_Fo_vs_Fg.min_length1000bp.txt\n",
      "links_Fo_vs_Fg.txt\n",
      "     circos/output :\n",
      "circos_Fg_chr1.vs.Fol4287_with_labels_and_colored_links.png\n",
      "circos_Fg_chr1.vs.Fol4287_with_labels_and_colored_links.svg\n",
      "circos_Fg_chr2.vs.Fol4287_with_labels_and_colored_links.png\n",
      "circos_Fg_chr2.vs.Fol4287_with_labels_and_colored_links.svg\n",
      "circos_Fg_chr3.vs.Fol4287_with_labels_and_colored_links.png\n",
      "circos_Fg_chr3.vs.Fol4287_with_labels_and_colored_links.svg\n",
      "circos_Fg_chr4.vs.Fol4287_with_labels_and_colored_links.png\n",
      "circos_Fg_chr4.vs.Fol4287_with_labels_and_colored_links.svg\n",
      "circos_karyotype_only.svg\n",
      "circos_karyotype_with_labels.png\n",
      "circos_karyotype_with_labels.svg\n",
      "circos_karyotype_with_labels_and_colored_links.png\n",
      "circos_karyotype_with_labels_and_colored_links.svg\n",
      "circos_karyotype_with_labels_and_links.png\n",
      "circos_karyotype_with_labels_and_links.svg\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "echo '     circos/input/config_files : '\n",
    "ls circos/input/config_files\n",
    "echo '     circos/input/karyotypes : '\n",
    "ls circos/input/karyotypes\n",
    "echo '     circos/input/links : '\n",
    "ls circos/input/links\n",
    "echo '     circos/output :'\n",
    "ls circos/output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "It's quite a lot isn't it?  \n",
    "Would you know in a few months from now what the difference is between all these files? Would you remember which is the last version and best version? \n",
    "\n",
    "For this problem, you can use a **versioning system**, that allows you to stack different versions of your files (code, a manuscript, a labjournal, protocols even) in an organized way, and where you can leave a sensible description of what is the difference between one version and the next. One system commonly used is Github. Give it a try! You can use for example this [guide](https://git-scm.com/book/en/v2) to get you started.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Progress in vizualization:  \n",
    "\n",
    "#### Dotplot, standard mummerplot output\n",
    "<p align=\"center\">\n",
    "<img src=\"out.png\" width=\"600px\" >\n",
    "</p>\n",
    "\n",
    "#### First Circosplot with links\n",
    "<p align=\"center\">\n",
    "<img src=\"circos_karyotype_with_labels_and_links.png\" width=\"600px\" >\n",
    "</p>\n",
    "\n",
    "#### Circosplot with _Fg_-colored links\n",
    "<p align=\"center\">\n",
    "<img src=\"circos_karyotype_with_labels_and_colored_links.png\" width=\"600px\" >\n",
    "</p>\n",
    "\n",
    "\n",
    "#### One circosplot with _Fo_-colored links per _Fg_ chromosome\n",
    "<img src=\"circos_Fg_chr1.vs.Fol4287_with_labels_and_colored_links.png\" style=\"float: left; width: 25%; margin-right: 0%;margin-bottom: 0.5em;\">\n",
    "<img src=\"circos_Fg_chr2.vs.Fol4287_with_labels_and_colored_links.png\" style=\"float: left; width: 25%; margin-right: 0%;margin-bottom: 0.5em;\">\n",
    "<img src=\"circos_Fg_chr3.vs.Fol4287_with_labels_and_colored_links.png\" style=\"float: left; width: 25%; margin-right: 0%;margin-bottom: 0.5em;\">\n",
    "<img src=\"circos_Fg_chr4.vs.Fol4287_with_labels_and_colored_links.png\" style=\"float: left; width: 25%; margin-right: 0%;margin-bottom: 0.5em;\">\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
