# Data about the data <a href='https://www.ebi.ac.uk/ena'><img src='../../img/ENA-logo.png' align="right" height="80" /></a>  
 - **Authors**: Like Fokkens (Postdoctorate fellow, SILS) & Marc Galland (Data Science/Manager Facilitator, SILS)
 - **Research topic**: Research Data Management
 - **Language**: None (Web browser)
 - **Level**: for beginners
 - **Lesson topic**: submitting your data and metadata to public repositories

## Learning objectives
After this lesson, you will:
- know the public repositories for genomic and metabolic datasets
- get an idea of what they expect in terms of data and metadata for submission
- understand the benefits of publishing datasets into public repositories

## Description
In this workshop we'll show some example of how metadata is used to retrieve relevant studies from online resources and we'll show some examples of metadata we've provided with studies we've submitted.

## Document your data: metadata
#### _How to use metadata from and provide metadata to online resources_    

### Searching for relevant datasets
Online databases contain a wealth of data generated by your fellow scientist. They submitted this data to allow others to reproduce any results they reported in a paper or to allow others to (re)analyse their data. The reusability of data strongly depends on how well this data is documented. To use data you need information on the experimental setup, where and how samples were collected, how the raw data was processed, etc.  

Online resources include:  

*  Genbank  	                 
A collection of databases, containing nucleotide and protein sequences, manuscripts (PubMed), projects, etc.  
* Sequencing Read Archive	  
DNA and RNA sequencing reads, from e.g. whole genome sequencing.  
* Gene Expression Omnibus      
DNA and RNA sequencing reads, from e.g. transcriptome sequencing or ChIP-seq experiments.  
* European Nucleotide Archive  
DNA and RNA sequencing reads, from e.g. whole genome or transcriptome sequencing or ChIP-seq experiments.  
* Zenodo
* Figshare

In this workshop we'll show you how you can search for specific datasets in one of these databases and show you examples of metadata
we've provided with data we've submitted ourselves to GEO and ENA.  

In this example we'll search for sequencing experiments in [sra]( https://www.ncbi.nlm.nih.gov/sra/ ). In the search bar you can type e.g. the name of
the organism that you're interested in. Under the search bar is a link [Advanced](https://www.ncbi.nlm.nih.gov/sra/advanced) in blue letters. If you click this link
you'll be directed to a page with the advanced search options available in SRA. These options are available because of the metadata provided by the data submitters.
Have a look at the fields in SRA. If you selected a field, you can check the option `Show index list`, this will show you the option you can choose.

For example if you want to select PacBio DNA sequencing reads in Fusarium oxysporum, you select:
	Field	Organism	"Fusarium oxysporum"
	Field	Platform	"pacbio smrt"
and select `Search`. Note that you could also type `fusarium oxysporum[Organism] AND "pacbio smrt"[Platform]` in the search bar directly: by selecting fields and values you are building a search.

Have a look at the results.  

If we go back to the `Advanced`-search page, please note that left from where you can select the field, you can select a logical operator to connects fields. By default this is `AND`. Have a look at the different operators.  

    Assignment:  
    Search for paired-end sequencing data in your favourite organism. Exclude whole genome sequencing studies.
    Select one of the results, have a close look at the record, do you recognise Fields in the record?


For more information you can read the [SRA search documentation](https://www.ncbi.nlm.nih.gov/sra/docs/srasearch/)

Now search for `(Fusarium oxysporum[Organism] AND "chip seq"[Strategy])`.  Select one of the records and study it. One of the attributes is a `GEO accession`: this record is cross-referenced between different databases.

Let's have a look this database that is also hosted by the NIH in the USA: the [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/gds/). This database is targeted towards functional genomics studies. Go to the top of the page, where you can select `GEO Datasets` in the menu left from the search bar. Select `Advanced` like you did in the SRA database before.  

    Assignment:
    Have a look the Fields. Note that they're different from what we saw at SRA!  
    Why do think this is?

The GEO is targeted towards experiments in which cases are compared to controls (functional genomics). That's why information on e.g. Sample size can be a very relevant criterium for selecting data that you might want to download. For a whole genome sequencing study, as we may find in SRA, this is not necessarily relevant.

### Providing metadata with your own dataset
Let's suppose we did ChIP-seq experiments on two histone modifications in _F. oxysporum_. We want to submit the sequencing data to an online database. But which one? You can check out NCBI's [Submission wizzard](https://www.ncbi.nlm.nih.gov/home/submit-wizard/).

GEO is probably the best option for this experiment. An added bonus is that we can also upload `.bam` files with ChIP-seq reads mapped to the genome and `.bed` files with regions that are enriched for a histone modification: we can also upload processed data that we drew conclusions from in our manuscript.

Let's go to [GEO](https://www.ncbi.nlm.nih.gov/geo/info/seq.html). You submission consists of:
  * Metadata spreadsheet  
  * Processed data files  
  * Raw data files  

After you've assembled all the files needed for the submission you can upload them via FTP.

For GEO, you provide the metadata in a [spreadsheet](https://www.ncbi.nlm.nih.gov/geo/info/seq.html#metadata). Let's download an example.

I used this example to assemble my own metadata sheet with my ChIP-seq experiments. Have a look.

    What do you think? Is there anything missing?

I've spent most time searching the labjournals of people actually performed the experiments and had already left the lab. They had properly documented their experiments, but it does take time to find this information, years after an experiment was performed.

    I strongly recommend keeping and updating the meta datafile as you
    generate and analyse your data (as sort of an extra labjournal).
    This will save you a lot of time in the end! Your metadata will be 
    more complete and therefor your dataset will have more value.

Another difficulty was that the example metadata spreadsheet all consider cases versus controls, which makes sense because GEO is for functional genomics. The template does not always easily fit your experimental setup.

    What can you do to solve this?

# The MetaboLights repository
From the [MetaboLights website](https://www.ebi.ac.uk/metabolights/):
> MetaboLights is a database for Metabolomics experiments and derived information. The database is cross-species, cross-technique and covers metabolite structures and their reference spectra as well as their biological roles, locations and concentrations, and experimental data from metabolic experiments.
MetaboLights is the recommended Metabolomics repository for a number of leading journals.

## Why would I use it?
- To store your own metabolomic research in a safe and reliable way.
- To get an accession number for your metabolomic study that you can cite in your publication (it is asked by reviewers).
- To be able to store the experimental _metadata_ along with the _raw data_.
- It is hosted by the European Bioinformatics Institute so it is a safe and reliable way to preserve and access your research.
- To collaborate with other research groups in an open but protected manner (authentification).

## What can I do with MetaboLights?
- Find metabolites and related metablomics studies by searching a wide range of associated metadata.
- Filter your search results on species, techniques and metabolites.
- Submit public or private studies.
- Receive a stable and unique accession number that can be used as a publication reference.
- Share private studies with collaborators/peer reviewers.
- Download public metabolomics studies for futher analysis.
- Retrieve molecular information from ChEBI or other linked compound databases.

## How to retrieve a dataset?
For instance, let's say your favorite organism is tomato :tomato: and you particularly like to work with a :de: research group headed by Alain Tissier. Just guessing!     
1. Go to the [MetaboLights website](https://www.ebi.ac.uk/metabolights/).
2. Under Study, click on "Browse" <img src='../../img/metabolights-browse.png' align="right" height="200" /></a>.
3. Filter for _study_ and _mass spectrometry_ in the __Type__ box.
4. Filter for _mass spectrometry_ in the __Technology__ box.
5. Type "Solanum lycopersicum" in the __Organism__ tool.
6. Finally, select "leaf" in the __Organism part__.

All submitted data is encoded in the standardised and open ISA-Tab format.

## Submitting a metabolic dataset to MetaboLights
The MetaboLights repository requires that experimental metadata are uploaded according to the _ISA-Tab_ file format.

### ISA and ISA-Tab formats
ISA stands for __"Investigation"__, __"Study"__ and __"Assay"__.  
* the __Investigation__ contains all the information needed to understand the overall goals and means used in an experiment.  
* the __Study__ and __Assay__ contains the experimental steps (or sequences of events).  
* For each __Investigation__ there may be one or more Study associated with it.  
* For each __Study__ there may be one or more __Assay__.  
<a href="https://isa-tools.org/format/specification.html"><img src='../../img/isa-model-structure.png' height="400"/></a>       

Description and goal of the ISA format:
> Built around the 'Investigation' (the project context), 'Study' (a unit of research) and 'Assay' (analytical measurement) data model and serializations (tabular, JSON and RDF), the ISA framework helps you to provide rich description of the experimental metadata (i.e. sample characteristics, technology and measurement types, sample-to-data relationships) so that the resulting data and discoveries are reproducible and reusable.  

The ISA-Tab is simply the tabulated version of this format.  

For instance, an __Investigation__ have to record the following information:

| Property   	|      Datatype      	|  Description 	|
|----------	|:-------------:	|------:	|
| Identifier 	|  String 	| A identifier or an accession number provided by a repository. This SHOULD be locally unique. 	|
| Title 	|    String   	|   A concise name given to the investigation. 	|
| Description 	| String 	|    A textual description of the investigation (Abstract of the paper for instance).  	|
| Submission Date 	|    Representation of a ISO8601 date   	|   The date on which the investigation was reported to the repository. 	|
| Public Release Date 	| Representation of a ISO8601 date 	|   The date on which the investigation was released publicly. 	|
| Publications 	|   A list of Publication   	|   A list of Publications relating to the investigation. 	|
| Contacts 	| A list of Contact 	|  A list of Contacts relating to the investigation.  	|

As another example, the __Assay__ should record the following information:

| Property   	|      Datatype      	|  Description 	|
|----------	|:-------------:	|------:	|
| Measurement type	|  Ontology annotation 	| An Ontology Annotation to qualify the endpoint, or what is being measured (e.g. gene expression profiling or protein identification). 	|
| Technology Type 	|    Ontology annotation   	|   An Ontology Annotation to identify the technology used to perform the measurement, e.g. DNA microarray, mass spectrometry. 	|
| Technology Platform 	| String 	|    The manufacturer and platform name, e.g. Bruker AVANCE, of the technology used. 	|



### ISAtools and ISAcreator
A toolbox called __isa-tools__ has been created to facilitate the handling and curation of experimental metadata.      
<a href="https://isa-tools.org/software-suite.html"><img src='../../img/isa-tools-website.png' height="600"/></a>     

Within that toolbox, you can use the ISAcreator to create the ISA-tabular format for MetaboLights.  
1. Download it [here](https://github.com/downloads/ISA-tools/ISAcreator/ISAcreator-1.5.zip) for all platforms.
2. Click on `ISAcreator.jar` to launch the graphical interface.
3. Then follow the steps described in the ISAcreator user manual: http://isatab.sourceforge.net/isahelp/ch04s02.html

### Submitting a dataset
1. Open the ISAcreator software and create a local user profile.
2. Create an ISArchive file that will contain the study, validate the experimental design and create an archive file.
3. Login and submit to MetaboLights.

A complete guide can be found here: https://www.ebi.ac.uk/training/online/course/metabolights-quick-tour/submitting-data-metabolights

# Useful links
* [Recommended data repositories by Nature Scientific Data](https://www.nature.com/sdata/policies/repositories).
* The European Nucleotide Archive: https://www.ebi.ac.uk/ena.
* The MetaboLights database: https://www.ebi.ac.uk/metabolights/.
* The Nature Genetics publication on the ISA format: https://www.nature.com/articles/ng.1054
